# Primera lista - Problema GRUPAL 1

Componentes del grupo:

*   Pol Casacuberta Gil
*   Joaquin Faraone Prieto
*   Agnes Felip i Díaz 
*   Marta Granero i Martí

## Dos mejor que una

### Consideremos un experimento en el que medimos una determinada variable aleatoria $X$, que sigue una distribución Gaussiana univariante $(X \sim \mathcal{N}(\mu,\,\sigma^{2}))$. Tomamos n medidas independientes de $X$ y obtenemos una muestra aleatoria $\{x_{1}, \dots, x_{n}\}$ donde cada $x_{i}$ es una realización de $X$, para $i = 1, \dots, n$. Resolved los siguentes apartados, ilustrando los resultados de la manera que os parezca más adecuada.


#### a) Escribid la función de densidad de probabilidad para una $x_{i}$ cualquiera y construid la función log-verosimilitud(negativa) de la muestra.

La expresion matemática que representa la función de densidad de una distribución normal para una $x_{i}$ es la siguiente:

$$
\begin{align}
\mathcal{N}(\textbf x_{i}| \mu, \sigma^2) = \frac{1}{\sqrt{2\pi\sigma^{2}}}exp(-\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}})
\end{align}
$$

Una vez dada la expresión de la función de densidad de la distribución normal para una $x_{i}$, nos piden construir la función de la log-verosimilitud negativa de la muestra.

Esta función de la log-verosimilitud($\mathcal{L}$) que vamos a construir, nos va a permitir encontrar los mejores parametros del modelo que se ajusten a los datos. O lo que es lo mismo, maximizaremos $\mathcal{L(\theta)}$ con respecto a los parámetros, $\theta$. 

Lo expresaremos como:

$$
\begin{align}
\theta = \underset{\theta}{argmax} \; p(\textbf X | \theta) 
\end{align}
$$

Si especificamos para nuestro caso, dónde p, sigue una distribución normal, tenemos:

$$
\begin{align}
\theta = \underset{\theta}{argmax} \; \mathcal{N}(\textbf X | \theta)
\end{align}
$$

Notemos pues que: $\mathcal{L(\theta)} = p(\textbf X | \theta) = \mathcal{N}(\textbf X | \theta)$

Como bien sabemos los parámetros para una distribución gaussiana, son $\mu$ y $\sigma^{2}$. Entonces para una MLE de un modelo gaussiano, solo hará falta encontrar buenas estimaciones de ambos parámetros usando derivadas parciales. 

Los parámetros $\theta$ son: $\mu, \sigma^{2}$:

$$
\begin{align}
\theta = \underset{\theta}{argmax} \; \mathcal{N}(\textbf X | \mu, \sigma^{2})
\end{align}
$$

Ahora sí, procedamos a construir la función:

Como tratamos con n muestras, las quales son (iid), los parámetros a estimar deberan tener en cuenta todo el conjunto de datos, por lo tanto sumaremos la log-verosimilitud para cada muestra de datos $i$ que tenemos. Lo expresamos como:

$$
\begin{align}
log (\mathcal{N}(\textbf X | \theta)) = \sum_{i=1}^{n} log (\mathcal{N}(\textbf x_{i} | \mu, \sigma^{2}))
\end{align}
$$

Ahora sustituyendo $\mathcal{N}(\textbf x_{i} | \mu, \sigma^{2})$, por su definición, nos queda:

$$
\begin{align}
\sum_{i=1}^{n} log (\mathcal{N}(\textbf x_{i} | \mu, \sigma^{2})) = \sum_{i=1}^{n} log (\frac{1}{\sqrt{2\pi\sigma^{2}}}exp(-\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}})
\end{align}
$$

Después de operar un poco tenemos la función de la log-verosimilitud. Ahora falta hacer-la negativa, así que simplemente negaremos la función obtenida anteriormente:

$$
\begin{align}
- \sum_{i=1}^{n} log (\frac{1}{\sqrt{2\pi\sigma^{2}}}exp(-\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}})
\end{align}
$$

Por el momento tendremos: 
$$
\begin{align}
\mathcal{L(\theta)} = - \sum_{i=1}^{n} log (\frac{1}{\sqrt{2\pi\sigma^{2}}}exp(-\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}})
\end{align}
$$

Podemos seguir simplificando y operando con esta función, hasta obtener:

$$
\begin{align}
\mathcal{L(\theta)} &= - \sum_{i=1}^{n} log (\frac{1}{\sqrt{2\pi\sigma^{2}}}exp(-\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}})\\
&= - \sum_{i=1}^{n} (log (\frac{1}{\sqrt{2\pi\sigma^{2}}}) + log (exp(-\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}}))) \\
&= - \sum_{i=1}^{n} ((log (1) - log (\sqrt{2\pi\sigma^{2}}) + log (\frac{-1}{2}(\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}})log(e))) \\
&= - \sum_{i=1}^{n} (- log (\sqrt{2\pi\sigma^{2}}) + log (\frac{-1}{2}(\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}}))) \\
&= - \sum_{i=1}^{n} (-\frac{1}{2} log (2\pi\sigma^{2}) - \frac{1}{2}(\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}})) \\
&= - \frac{n}{2} \; log (2\pi\sigma^{2}) + \sum_{i=1}^{n} - \frac{1}{2}(\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}}) \\
&= - \frac{n}{2} \; log (2\pi\sigma^{2}) - \frac{1}{2\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2} \\
\end{align} 
$$

En esta última función obtenida, lo que nos interesará minimizar la pérdida cuadrática que tenemos en el segundo término. 

#### b) Encontrad los estimadores de máxima verosimilitud $\hat \mu$ y $\hat \sigma^{2}$, a partir de la muestra.

Ahora que tenemos la construcción de la función de la log-verosimilitud negativa de la muestra, podemos encontrar los estimadores de máxima verosimilitud $\hat \mu$ y $\hat \sigma^{2}$ con ayuda de las derivadas parciales.

Así pues mediante estas queremos encontrar el mejor valor para cada uno. Formalizado como:
$$
\begin{align}
\underset{\mu}{argmax} \; \mathcal{L}(\textbf X | \mu, \sigma^{2}) \\
\underset{\sigma^{2}}{argmax} \; \mathcal{L}(\textbf X | \mu, \sigma^{2}) \\
\end{align}
$$

Entonces, para maximizar $\mathcal{L(\theta)}$ con respecto a los parámetros, $\mu$ y $\sigma^{2}$ debemos la derivada parcial de la función con respecto al parámetro que consideremos, establecer esa derivada parcial en cero, y resolver para el parámetro. Es decir:

$$
\begin{align}
\underset{\mu}{argmax} \; \mathcal{L}(\textbf X | \mu, \sigma^{2})  = \frac{\partial\mathcal{L}}{\partial\mu} := 0\\
\underset{\mu}{argmax} \; \mathcal{L}(\textbf X | \mu, \sigma^{2})  = \frac{\partial\mathcal{L}}{\partial\sigma^{2}} := 0 \\
\end{align}
$$

Empezamos con la derivada parcial en $\mu$:
$$
\begin{align}
\frac{\partial\mathcal{L}}{\partial\mu} &= \frac{\partial}{\partial\mu}(- \frac{n}{2} \; log (2\pi\sigma^{2}) - \frac{1}{2\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
&= \frac{\partial}{\partial\mu}(- \frac{n}{2} \; log (2\pi\sigma^{2})) + \frac{\partial}{\partial\mu} (- \frac{1}{2\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
&= 0 + \frac{\partial}{\partial\mu} (- \frac{1}{2\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
&= \frac{\partial}{\partial\mu} (- \frac{1}{2\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
&= \frac{\partial}{\partial\mu} (\sum_{i=1}^{n} - \frac{1}{2\sigma^{2}} (x_{i} - \mu)^{2}) \\
&= \sum_{i=1}^{n} \frac{\partial}{\partial\mu} (- \frac{1}{2\sigma^{2}} (x_{i} - \mu)^{2}) \\
&= \sum_{i=1}^{n} (\frac{\partial}{\partial\mu} (- \frac{1}{2\sigma^{2}}) (x_{i} - \mu)^{2} + (-\frac{1}{2\sigma^{2}}) \frac{\partial}{\partial\mu}(x_{i} - \mu)^{2}) \\
&= \sum_{i=1}^{n} (0 + (-\frac{1}{2\sigma^{2}}) \frac{\partial}{\partial\mu}(x_{i} - \mu)^{2}) \\
&= -\frac{1}{2\sigma^{2}} \sum_{i=1}^{n} \frac{\partial}{\partial\mu}(x_{i} - \mu)^{2} \\
&= -\frac{1}{2\sigma^{2}} \sum_{i=1}^{n} -2(x_{i} - \mu) \\
&= \frac{1}{\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu) \\
\end{align}
$$

Ahora, tenemos la forma más simple de la derivada parcial de nuestra función de verosimilitud con respecto a $\mu$, así que igualamos a 0 para encontrar el mejor parámetro $\mu$.

$$
\begin{align}
\frac{\partial\mathcal{L}}{\partial\mu} &= \frac{1}{\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu) \\
0 &= \frac{1}{\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu) \\
0 &= \sum_{i=1}^{n} (x_{i} - \mu) \\
0 &= \sum_{i=1}^{n} x_{i} - \sum_{i=1}^{n} \mu \\
0 &= \sum_{i=1}^{n} x_{i} - n \mu \\
n \mu &= \sum_{i=1}^{n} x_{i} \\
\mu &= \frac{1}{n} \sum_{i=1}^{n} x_{i} \\
\end{align}
$$

Por lo tanto el mejor estimador para $\mu$ será: 

$$
\begin{align}
\hat \mu = \frac{1}{n} \sum_{i=1}^{n} x_{i}
\end{align}
$$

Si repetimos el proceso para el parámetro $\sigma^{2}$, tendremos:
$$
\begin{align}
\frac{\partial\mathcal{L}}{\partial\sigma^{2}} &= \frac{\partial}{\partial\sigma^{2}}(- \frac{n}{2} \; log (2\pi\sigma^{2}) - \frac{1}{2\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
&= \dots \\ 
&= \frac{-n}{2} \frac{\partial}{\partial\sigma^{2}}(log (2\pi\sigma^{2})) + \frac{\partial}{\partial\sigma^{2}}(- \frac{1}{2\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
&= \dots \\
&= \frac{-n}{2\sigma^{2}} + \frac{\partial}{\partial\sigma^{2}}(- \frac{1}{2\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
&= \frac{-n}{2\sigma^{2}} + \sum_{i=1}^{n} (\frac{\partial}{\partial\sigma^{2}}(- \frac{1}{2\sigma^{2}} (x_{i} - \mu)^{2})) \\
&= \dots \\
&= \frac{-n}{2\sigma^{2}} + \sum_{i=1}^{n} (- \frac{1}{2\sigma^{2}} \frac{\partial}{\partial\sigma^{2}} \sigma^{-2} (x_{i} - \mu)^{2}))  \\
&= \dots \\
&= \frac{1}{2\sigma^{2}}(-n + \frac{1}{\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
\end{align}
$$

Ahora, tenemos la forma más simple de la derivada parcial de nuestra función de verosimilitud con respecto a $\sigma^{2}$, así que igualamos a 0 para encontrar el mejor parámetro $\sigma^{2}$.

$$
\begin{align}
\frac{\partial\mathcal{L}}{\partial\sigma^{2}} &= \frac{1}{2\sigma^{2}}(-n + \frac{1}{\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2}) \\
0 &= -n + \frac{1}{\sigma^{2}} \sum_{i=1}^{n} (x_{i} - \mu)^{2} \\
n \sigma^{2} &= \sum_{i=1}^{n} (x_{i} - \mu)^{2} \\
\sigma^{2} &= \frac{1}{n} \sum_{i=1}^{n} (x_{i} - \mu)^{2} \\
\end{align}
$$

Por lo tanto el mejor estimador para $\sigma^{2}$ será: 

$$
\begin{align}
\hat \sigma^{2} = \frac{1}{n} \sum_{i=1}^{n} (x_{i} - \mu)^{2}
\end{align}
$$

#### c) Demostrad que realmente son minimos(y no extremos cualquiera).

Para poder demostrar que ambos estimadores: $\hat \sigma^{2}$ y $\mu$ son mínimos nos seguiremos ayudando de las derivadas parciales. Tomaremos para cada parámetro, sus segundas derivadas parciales y estudiaremos la concavidad de la función log-verosimiltud. 

Para eso, bastará ver que ambas derivadas segundas son positivas para poder afirmar que ambos parámetros son mínimos. Formalizado como:

$$
\begin{align}
\frac{\partial^{2}\mathcal{L}}{\partial^{2}\sigma^{2}} > 0 \; ; \frac{\partial^{2}\mathcal{L}}{\partial^{2}\mu} > 0 \\
\end{align}
$$

Para no hacer pesada la lectura del apartado, daremos las expresiones analíticas directamente y justificaremos porque las derivadas son positivas.

Para el parámetro $\mu$:

$$
\begin{align}
\frac{\partial^{2}\mathcal{L}}{\partial^{2}\mu^{2}} &= n\frac{1}{\sigma^{2}} \\
\end{align}
$$

Y para el parámetro $\sigma^{2}$:

$$
\begin{align}
\frac{\partial^{2}\mathcal{L}}{\partial^{2}\sigma^{2}} = \frac{1}{\frac{1}{\sigma^{2}}} &= \sigma^{2} \\
\end{align}
$$

Como podemos observar ambas derivadas parciales segundas con respecto al parámetro $\mu$ y al parámetro $\sigma^{2}$ son positivas. Estas seran positivas en todo su dominio, ya que tanto $n$ como $\sigma^{2}$ siempre tomaran valores positivos, por lo tanto, no solo son extremos, sino que son mínimos.

Formalizado para el parámetro $\mu$ y $\sigma^{2}$:

$$
\begin{align}
\frac{\partial^{2}\mathcal{L}}{\partial^{2}\mu^{2}} = n\frac{1}{\sigma^{2}} > 0; \\
\end{align}
$$

$$
\begin{align}
\frac{\partial^{2}\mathcal{L}}{\partial^{2}\sigma^{2}} = \frac{1}{\frac{1}{\sigma^{2}}} = \sigma^{2} > 0\\
\end{align}
$$

Por lo tanto, podemo afirmar que la función negativa de la log-verosimilitud para la distribución gaussiana es no concava, y por lo tanto convexa para el parámetro que representa la media y para el parámetro de la varianza.

Notamos, que la función de la log-verosimilitud será concava para ambos parámetros.

#### d) Implementad la función de log verosimilitud usand JAX. Definid muestras (independientes) de tamaño 50, 500, 5000 para una distribución Gaussiana com $\mu$ y $\sigma$ de vuestra elección. Implementad un algoritmo de descenso del gradiente utilizando JAX para optimizar la log verosimilitud y estimar los parámetros de la distribución explorandola tasa de aprendizaje y el número máximo de iteraciones. Podéis emplear un valor $\epsilon$ para acabr la optimización de $1e-10$ e inicializar los vallores de los parámetrosa un valor razonable que no esté ni demasiado lejos, ni demasiado cerca de los valores reales. Comparad el resultado con el cálculo que dan los estimadores teóricos y comentad el resultado

In [ ]:
import jax.numpy as jnp
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats
import numpy as np
from jax import grad, vmap, jit, random, lax
from numpy.random import seed
from numpy.random import normal
from jax.scipy.stats import norm

rng = np.random.default_rng(42) # semilla para fijar la experimentación

#Valores reales
sigma = 3
media = 10

#(iid) de tamaño 50
data1 = rng.normal(loc=media, scale=sigma, size=50)
#(iid) de tamaño 500
data2 = rng.normal(loc=media, scale=sigma, size=500)
#(iid) de tamaño 5000
data3 = rng.normal(loc=media, scale=sigma, size=5000)

print ("Parámetros de las muestras:")
var1 = jnp.std(data1, ddof=1)
var2 = jnp.std(data2, ddof=1)
var3 = jnp.std(data3, ddof=1)

print ("Sigma1: ", var1)
print ("Sigma2: ",var2)
print ("Sigma3: ",var3)

media1 = jnp.mean(data1)
media2 = jnp.mean(data2)
media3 = jnp.mean(data3)
print ("Media1: ", media1)
print ("Media2: ", media2)
print ("Media3: ", media3)

def estimate_mu_sigma(data):
    epsilon = 1e-6
    lr = 0.5
    n_iter = 10000

    #Valores que suponemos inicialmente
    mu = 15.0
    sigma = 10.0


    grad_likelohood_mu = jit(grad(likelohood, argnums = 0)) 
    grad_likelohood_sigma = jit(grad(likelohood, argnums = 1))

    grad_mu_mean = 0
    grad_sigma_mean = 0  
    for i in range(n_iter):
        gradiente_mu = vmap(grad_likelohood_mu,in_axes=(None, None, 0))(mu, sigma, data)
        gradiente_sigma = vmap(grad_likelohood_sigma,in_axes=(None, None, 0))(mu, sigma, data)

        grad_mu_mean = jnp.mean(gradiente_mu)
        grad_sigma_mean = jnp.mean(gradiente_sigma)

        mu = mu - (lr * grad_mu_mean)
        sigma = sigma - (lr * grad_sigma_mean)
        if grad_mu_mean < epsilon and grad_sigma_mean < epsilon: 
            print("hemos llegado al minimo")
            print(i)
            break
    print("sigma: " + str(sigma), "mu: " + str(mu))

def likelohood(mu, sigma, data):
    n = data.size
    ll = (1/(2*sigma**2)) *  jnp.sum((data-mu)**2) + (n/2)*jnp.log(sigma**2) + (n/2)*jnp.log(2*jnp.pi)
    return ll

print("data1:")
estimate_mu_sigma(data1)
print("data2:")
estimate_mu_sigma(data2)
print("data3:")
estimate_mu_sigma(data3)

Parámetros de las muestras:
Sigma1:  2.3048317
Sigma2:  2.9645505
Sigma3:  3.0020196
Media1:  10.273633
Media2:  9.919176
Media3:  9.950218
data1:
hemos llegado al minimo
254
sigma: 2.2816672 mu: 10.273638
data2:
hemos llegado al minimo
343
sigma: 2.961585 mu: 9.919183
data3:
hemos llegado al minimo
349
sigma: 3.0017204 mu: 9.950227


#### Observaciones:
Hemos implementado un algoritmo de descenso de gradiente y jugando un poco con los paràmetros learning rate $lr$ y con el número máximo de iteraciones. Parece que con un learning rate de 0.5 y unas 10.000 iteraciones se consiguen resultados muy cercanos a los estimadores teóricos de $\mu$ y $\sigma^{2}$.

e) No todo se comporta según una distribución Gaussiana. A veces nuestros datos son generados por varios procesos diferentes que siguen distribuciones Gausianas con parámetros diferentes, básicamente tenemos una distribución conjunta con varias modalidades. Escribid la función de densidad de probabilidad para una distribución que sigue la suma de k Gaussianas (fijaos que ha de ser una distribución de probabilidad). Supongamos que tenemos dos máquinas que fabrican dos tipos de tornillos de distinta longitud con cierta varianza también distinta. Todos los tornillos acaban mezclados y queremos calcular cuáles son los parámetros de la distribución de cada máquina. Sabemos que cada máquina ha fabricado el mismo número de tornillos. Generad dos muestras (independientes) de igual tamaño (2500) para cada distribución usando el generador de muestras normales de JAX. Mirad como se trata la generación de muestras en el capítulo 3 para que sean independientes. Implementad un algoritmo de descenso de gradiente para optimizar los parámetros de las dos distribuciones con la muestra conjunta utilizando la función de log verosimilitud de la mezcla de dos gaussianas. Explorad diferentes valores para la tasa de aprendizaje y el número máximo de iteraciones.

Assumiendo que las muestras son independientes entonces sabemos que su suma también se distribuye de forma nomal: 

$$ X\sim N(\mu _{X},\sigma _{X}^{2})$$
$$ Y\sim N(\mu _{Y},\sigma _{Y}^{2})$$
$$ Z = X + Y$$

Con lo que equivaldrá a:

$$ Z \sim (\mu _{Y}+\mu _{X},\sigma _{X}^{2}+\sigma _{Y}^{2}) $$


Este ejemplo con dos variables nos lleva a: 

$$ K \sim \sum \limits_{i=1}^{n}(\mu _{i},\sigma_{i}^{2}) $$


Teniendo en cuenta que la funcion de distribución es:


$$
\begin{align}
\mathcal{f}(\textbf x_{i}| \mu, \sigma^2) = \frac{1}{\sqrt{2\pi\sigma^{2}}}exp(-\frac{(x_{i} - \mu)^{2}}{2\sigma^{2}})
\end{align}
$$

Si aplicamos esta propiedad a la función de distribución usando unas muestras para $k$ distribuciones gaussianas de diferentes parametros: $\vec{\mu} = \{ \mu_1, ..., \mu_k \}$ i $\vec{\sigma} = \{ \sigma_1, ..., \sigma_k \}$ pero generalmente la mediana y varianza de la suma de K variables independientes distribuidas de forma identica se da por $\mu_K=K\mu$ y $\sigma_K=K\sigma^{2}$ con lo que la funcion de distribucion resultante es:

$$
\begin{align}
\mathcal{f}(\textbf x_{i}| \mu, \sigma) = \frac{1}{\sqrt{2\pi K\sigma^{2}}}exp(-\frac{(x_{i} - \mu_K)^{2}}{2\sigma_K^{2}})
\end{align}
$$



In [3]:
import jax.numpy as jnp
import math
from jax import grad, vmap, jit, random
from scipy import stats
rng = np.random.default_rng(42)
mu = 10
sigma = 3
data1 = rng.normal(loc=mu, scale=sigma, size=2500)

mu2 = 15
sigma2 = 2
data2 = rng.normal(loc=mu2, scale=sigma2, size=2500)

data = np.concatenate((data1, data2))

def estimate_mu_sigma_E(data):
    epsilon = 1e-10
    lr = 0.5
    mu = [12.0, 2.0]
    sigma = [16.0, 4.0]
    grad_likelohood_mu = jit(grad(likelihood_two, argnums=0))
    grad_likelohood_sigma = jit(grad(likelihood_two, argnums=1))
    grad_mu_mean = [0, 0] 
    grad_sigma_mean = [0, 0]
    for i in range(10000):
        gradiente_mu = vmap(grad_likelohood_mu, in_axes=(
            None, None, 0))(mu, sigma, data)
        gradiente_sigma = vmap(grad_likelohood_sigma,
                               in_axes=(None, None, 0))(mu, sigma, data)
        grad_mu_mean[0] = jnp.mean(gradiente_mu[0])
        grad_mu_mean[1] = jnp.mean(gradiente_mu[1])
        grad_sigma_mean[0] = jnp.mean(gradiente_sigma[0])
        grad_sigma_mean[1] = jnp.mean(gradiente_sigma[1])
        mu[0] = mu[0] - (lr * grad_mu_mean[0])
        mu[1] = mu[1] - (lr * grad_mu_mean[1])
        sigma[0] = sigma[0] - (lr * grad_sigma_mean[0])
        sigma[1] = sigma[1] - (lr * grad_sigma_mean[1])
        if grad_mu_mean[0] < epsilon and grad_sigma_mean[0] < epsilon and grad_mu_mean[1] < epsilon and grad_sigma_mean[1] < epsilon:
            print("hemos llegado al minimo")
            print(i)
            break
    print("mu 1: " + str(mu[0]))
    print("sigma 1: " + str(sigma[0]))
    print("mu 2: " + str(mu[1]))
    print("sigma 2: " + str(sigma[1]))


def likelihood_two(mu, sigma, data):
    n = data.size
    ll = (1/(2*(sigma[0]**2+sigma[1]**2))) * jnp.sum((data-(mu[0] + mu[1]))**2) + (n/2)*jnp.log((sigma[0]**2+ sigma[1]**2)) + (n/2)*jnp.log(2*jnp.pi)
    return ll

estimate_mu_sigma_E(data)

mu 1: 11.219053
sigma 1: 3.515402
mu 2: 1.2190298
sigma 2: 0.8788505


#### Observaciones:

Hemos intentado utilizar JAX para encontrar los parámetros de las muestras de dos gaussianas distintas, aprovechandonos de la capacidad de la libreria para calcular gradientes sobre un vector de parámetros. Sin embargo no parece que el segundo parámetro de mu i de sigma se aproxime correctamente, aunque el primer parámetro de estos sí se acerca a los valores deseados.

#### f) Emplear la misma tasa de aprendizaje durante toda la optimización puede evitar que lleguemos realmente cerca del óptimo. Modificad el algoritmo que habéis implementado para que la tasa de aprendizaje vaya atenuándose con el número de iteraciones multiplicándola por un valor cercano a 1 $(0,9999, 0.999, 0,99, \dots, 0.9)$. Repetid la optimización de vuestro mejor resultado en el apartado anterior con diferentes atenuaciones. ¿Ha afectado al resultado? ¿Ha afectado al número de iteraciones antes de converger?

In [ ]:
def estimate_mu_sigma_dec(data):
    epsilon = 1e-6
    lr = 0.5
    mu = 15.0
    sigma = 10.0
    grad_likelohood_mu = jit(grad(likelohood, argnums = 0)) 
    grad_likelohood_sigma = jit(grad(likelohood, argnums = 1))
    grad_mu_mean = 0
    grad_sigma_mean = 0
    iterate = 10000
    dec = np.linspace(1, 0.1, iterate)
# dec = np.linspace(0.0002, 0.001, iterate)
    # print(dec)
    for i in range(iterate):
        gradiente_mu = vmap(grad_likelohood_mu,in_axes=(None, None, 0))(mu, sigma, data)
        gradiente_sigma = vmap(grad_likelohood_sigma,in_axes=(None, None, 0))(mu, sigma, data)
        grad_mu_mean = jnp.mean(gradiente_mu)
        grad_sigma_mean = jnp.mean(gradiente_sigma)
        mu = mu - (lr * grad_mu_mean*dec[i])
        sigma = sigma - (lr * grad_sigma_mean*dec[i])
        if grad_mu_mean < epsilon and grad_sigma_mean < epsilon: 
            print("hemos llegado al minimo")
            print(i)
            break
    print("sigma: " + str(sigma))
    print("mu: " + str(mu))


print("data1")
estimate_mu_sigma_dec(data1)
print("data2")
estimate_mu_sigma_dec(data2)
print("data3")
estimate_mu_sigma_dec(data3)

data1
hemos llegado al minimo
257
sigma: 2.2816675
mu: 10.273638
data2
hemos llegado al minimo
349
sigma: 2.961585
mu: 9.919184
data3
hemos llegado al minimo
354
sigma: 3.0017204
mu: 9.950227


#### Observaciones:



Empleando distintas atenuaciones podemos ver que el número de iteraciones que tarda en converger es mayor pero tampoco de forma muy significativa (por ej. de 254 iteraciones a 257 en el primer conjunto de datos). Con atenuaciones más agresivas, se pueden conseguir resultados no deseables como que no podamos llegar a converger ya que el learning rate llegue a ser demasiado pequeño.